domains/pallets/auto-id/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+        }
+    }
+
+    fn issue_certificate_serial<T: Config>(&mut self, serial: U256) -> DispatchResult {
+        match self {
+            Certificate::X509(cert) => {
+                ensure!(
+                    !cert.issued_serials.contains(&serial),
+                    Error::<T>::CertificateSerialAlreadyIssued
+                );
+                cert.issued_serials.insert(serial);
+                Ok(())
+            }
+        }
+    }
+
+    /// Checks if the certificate is valid at this time.
+    pub(crate) fn is_valid_at(&self, time: Moment) -> bool {
+        match self {
+            Certificate::X509(cert) => cert.validity.is_valid_at(time),
+        }
+    }
+
+    fn revoke(&mut self) {
+        match self {
+            Certificate::X509(cert) => cert.revoked = true,
+        }
+    }
+
+    fn is_revoked(&self) -> bool {
+        match self {
+            Certificate::X509(cert) => cert.revoked,
+        }
+    }
+}
+
+/// A representation of AutoId
+#[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
+pub struct AutoId {
+    /// Unique AutoID identifier.
+    pub identifier: Identifier,
+    /// Certificate associated with this AutoId.
+    pub certificate: Certificate,
+}
+
+/// Type holds X509 certificate details used to register an AutoId.
+#[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
+pub enum RegisterAutoIdX509 {
+    Root {
+        certificate: DerVec,
+        signature_algorithm: DerVec,
+        signature: Vec<u8>,
+    },
+    Leaf {
+        issuer_id: Identifier,
+        certificate: DerVec,
+        signature_algorithm: DerVec,
+        signature: Vec<u8>,
+    },
+}
+
+/// Signature holds algorithm used and the signature value.
+#[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
+pub struct Signature {
+    pub signature_algorithm: DerVec,
+    pub value: Vec<u8>,
+}
+
+/// Request to register a new AutoId.
+#[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
+pub enum RegisterAutoId {
+    X509(RegisterAutoIdX509),
+}
+
+#[frame_support::pallet]
+mod pallet {
+    use crate::{AutoId, Identifier, RegisterAutoId, Signature};
+    use frame_support::pallet_prelude::*;
+    use frame_support::traits::Time;
+    use frame_system::pallet_prelude::*;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+        type Time: Time<Moment = subspace_runtime_primitives::Moment>;
+    }
+
+    #[pallet::pallet]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(_);
+
+    /// Stores the next auto id identifier.
+    #[pallet::storage]
+    pub(super) type NextAutoIdIdentifier<T> = StorageValue<_, Identifier, ValueQuery>;
+
+    /// Stores the auto id identifier against an AutoId.
+    #[pallet::storage]
+    pub(super) type AutoIds<T> = StorageMap<_, Identity, Identifier, AutoId, OptionQuery>;
```

Comment:
I'm not sure at this moment. Some of the structures would change as we iterate this pallet. I would not want to remove and add this struct back since AutoId necessarily does not hold only certificate IIUC. I'll revisit this soon